### ISSUE TITLE
Mac + Reticulate + RStudio 2023.09.0 or later

### ISSUE THREAD

**Issue by anirban-mu** [rstudio/rstudio#13967]
### System details

    RStudio Edition : Desktop
    RStudio Version : RStudio 2023.09.0 / 2023.09.1
    OS Version      : Mac OS 14.1.1
    R Version       : 2023-11-20 r85569

### Steps to reproduce the problem

Python is installed via home brew. The path is passed to Reticulate as follows.

    Sys.setenv(RETICULATE_PYTHON = "/opt/Homebrew/Cellar/python@3.11/3.11.6_1/bin/python3")
    Sys.setenv(TENSORFLOW_PYTHON = "/opt/Homebrew/Cellar/python@3.11/3.11.6_1/bin/python3")
    library(reticulate)

In RStudio versions prior to Desert Sunflower, when I import umap, it works as expected. 

    > reticulate::repl_python()
    Python 3.11.6 (/opt/homebrew/Cellar/python@3.11/3.11.6_1/bin/python3)
    Reticulate 1.34.0 REPL -- A Python interpreter in R.
    Enter 'exit' or 'quit' to exit the REPL and return to R.
    >>> import numpy as np
    >>> import pandas as pd
    >>> import umap.umap_ as umap

No issues here.

### Describe the problem in detail

In 2023.09.0 and later (including the current version -- 2023.09.1), the correct Python is picked up but then I get an error:

    > reticulate::repl_python()
    Python 3.11.6 (/opt/homebrew/Cellar/python@3.11/3.11.6_1/bin/python3)
    Reticulate 1.34.0 REPL -- A Python interpreter in R.
    Enter 'exit' or 'quit' to exit the REPL and return to R.
    >>> import numpy as np
    >>> import pandas as pd
    >>> import umap.umap_ as umap
    Traceback (most recent call last):
      File "<python_lib_path>/site-packages/llvmlite/binding/ffi.py", line 134, in __getattr__
        return self._fntab[name]
               ~~~~~~~~~~~^^^^^^
    KeyError: 'LLVMPY_AddSymbol'
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "<python_lib_path>/site-packages/llvmlite/binding/ffi.py", line 115, in _load_lib
        self._lib_handle = ctypes.CDLL(str(lib_path))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "<python_lib_path>/lib/python3.11/ctypes/__init__.py", line 376, in __init__
        self._handle = _dlopen(self._name, mode)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^
    OSError: dlopen(<python_lib_path>/site-packages/llvmlite/binding/libllvmlite.dylib, 0x0006): Library not loaded: @rpath/libz.1.dylib
      Referenced from: <dynamic_library_path> /site-packages/llvmlite/binding/libllvmlite.dylib
      Reason: tried: '<other_paths>'
    
    The above exception was the direct cause of the following exception:
    
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "<python_lib_path>/site-packages/umap/umap_.py", line 29, in <module>
        import numba
      File "<r_library_path>/python/rpytools/loader.py", line 119, in _find_and_load_hook
        return _run_hook(name, _hook)
               ^^^^^^^^^^^^^^^^^^^^^^
      ... [repeated stack trace omitted for brevity]
      File "<python_lib_path>/site-packages/llvmlite/binding/__init__.py", line 4, in <module>
        from .dylib import *
      File "<r_library_path>/python/rpytools/loader.py", line 119, in _find_and_load_hook
        return _run_hook(name, _hook)
               ^^^^^^^^^^^^^^^^^^^^^^
      ... [repeated stack trace omitted for brevity]
      File "<python_lib_path>/site-packages/llvmlite/binding/dylib.py", line 36, in <module>
        ffi.lib.LLVMPY_AddSymbol.argtypes = [
        ^^^^^^^^^^^^^^^^^^^^^^^^
      File "<python_lib_path>/site-packages/llvmlite/binding/ffi.py", line 137, in __getattr__
        cfn = getattr(self._lib, name)
                      ^^^^^^^^^
      File "<python_lib_path>/site-packages/llvmlite/binding/ffi.py", line 129, in _lib
        self._load_lib()
      File "<python_lib_path>/site-packages/llvmlite/binding/ffi.py", line 123, in _load_lib
        raise OSError("Could not find/load shared object file") from e
    OSError: Could not find/load shared object file

I trimmed the output to anonymise but I hope the details are clear.

I installed and uninstalled RStudio leaving all other libraries (including R, Python, and umap the same) and I was able to replicate this behavior whereby there is no error with prior versions of RStudio and an error with a newer version of RStudio. In addition, if I launch Python through terminal, I do not get an error. Finally, many other Python modules load ok but umap always leads to the same error.

All of this suggests that some change in Desert Sunflower broke the configuration in my system. I suspect this has to do with the change in the handling of `DYLD_FALLBACK_LIBRARY_PATH`.

I have installed and reinstalled reticulate and all other R and Python libraries but nothing seems to help. It is only when I revert to the old version of RStudio that the error goes away.

I also get this error if I try to import numba when running Python through Reticulate in the newer version of RStudio but not otherwise. As numba is used by umap, I suspect that this may be interrelated or even the source of the error. However, I am clueless beyond that.

My system is a bit cruddy and has old libraries. Something could have been left behind that is causing the issue or it may be the path. However, the fact that the older version of RStudio and running Python directly on Terminal gives me the expected behavior makes me pretty sure some change in the new version is key to this.

If this is replicable then the following steps "should" replicate it.

1. Install Python 3.11 using home brew.
2. Install umap-learn using pip.
3. Set Python path, run the interpreter, and load the library.
4. Import umap or numba.

### Describe the behavior you expected

Reticulate + Python in newer versions of RStudio should be the same as Python on terminal and Reticulate + Python on older versions of RStudio.

For now, I am going to simply use an older version of RStudio but I am happy to help debug if anyone has specific questions. I must admit I am unsure of where to start as its quite a specific error, but I am happy to report specifics from both prior and new versions if you tell me what specifics to report.

<!-- 
Please keep the below portion in your issue, and check `[x]` the applicable boxes.
-->

- [X ] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [X ] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ X] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.posit.co/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [X ] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by anirban-mu** [comment]
SIP is enabled so `DYLD_FALLBACK_LIBRARY_PATH` should not matter/work.

```
csrutil status
System Integrity Protection status: enabled.
```
`echo $DYLD_LIBRARY_PATH` in terminal gives me the empty string, which is as expected. In the prior version of RStudio I get:

```
> Sys.getenv("DYLD_FALLBACK_LIBRARY_PATH")
[1] ""
```
In the new version of Rstudio I get: 

```
> Sys.getenv("DYLD_FALLBACK_LIBRARY_PATH")
[1] "/Library/Frameworks/R.framework/Resources/lib:/opt/homebrew/Cellar/openjdk/21.0.1/libexec/openjdk.jdk/Contents/Home/lib/server:/var/folders/qd/7j99j7f123l170w8h2dr6yr80000gn/T/rstudio-fallback-library-path-1673374761"
```

I think this might be the issue. I am unsure of where RStudio is picking this string up from. I can see that there was a change made to how RStudio forwards this variable. Maybe the test to see if SIP is on is somehow passing on my system and it's picking up something from homebrew? Regardless, is SIP is on then this variable should be empty AFAIK.

**Comment by kevinushey** [comment]
FWIW that's kind of surprising to me, because on macOS R should normally place its own library directory on the fallback library path:

```
$ R -s -e 'Sys.getenv("DYLD_FALLBACK_LIBRARY_PATH")'
[1] "/Library/Frameworks/R.framework/Resources/lib:/Library/Java/JavaVirtualMachines/jdk-11.0.18+10/Contents/Home/lib/server"
```

And this all happens in the R installation's `ldpaths` script:

```
kevin@MBP-P2MQ:/Library/Frameworks/R.framework/Resources/etc
$ cat ldpaths
: ${JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk-11.0.18+10/Contents/Home}
: ${R_JAVA_LD_LIBRARY_PATH=${JAVA_HOME}/lib/server}
if test -n ""; then
: ${R_LD_LIBRARY_PATH=${R_HOME}/lib:}
else
: ${R_LD_LIBRARY_PATH=${R_HOME}/lib}
fi
if test -n "${R_JAVA_LD_LIBRARY_PATH}"; then
  R_LD_LIBRARY_PATH="${R_LD_LIBRARY_PATH}:${R_JAVA_LD_LIBRARY_PATH}"
fi
## This is DYLD_FALLBACK_LIBRARY_PATH on Darwin (macOS) and
## LD_LIBRARY_PATH elsewhere.
## However, on macOS >=10.11 (if SIP is enabled, the default), the
## environment value will not be passed to a script such as R.sh, so
## would not seen here.
if test -z "${DYLD_FALLBACK_LIBRARY_PATH}"; then
  DYLD_FALLBACK_LIBRARY_PATH="${R_LD_LIBRARY_PATH}"
else
  DYLD_FALLBACK_LIBRARY_PATH="${R_LD_LIBRARY_PATH}:${DYLD_FALLBACK_LIBRARY_PATH}"
fi
export DYLD_FALLBACK_LIBRARY_PATH
```

Note that the Java components would be toggled based on the result of `sudo R CMD javareconf`.

Finally, I can also confirm I do have SIP enabled, e.g.

```
$ csrutil status
System Integrity Protection status: enabled.
```

Finally, in case it matters, I see the same sort of thing with Homebrew R, e.g.

```
$ /opt/homebrew/bin/R -s -e 'Sys.getenv("DYLD_FALLBACK_LIBRARY_PATH")'
[1] "/opt/homebrew/Cellar/r/4.3.2/lib/R/lib"
```

**Comment by anirban-mu** [comment]
Just to confirm, I get the same error in Ocean Storm. Not trying to rush anyone. Just confirming that I don't see a change.

**Comment by anirban-mu** [comment]
I have managed to isolate the issue to `import llvmlite.binding`. I will submit a bug report to llvmlite.

It's an oddly specific issue given that it works fine on Python shell. It's only when either using Reticulate + R through terminal or in the newer version of RStudio that the bug manifests.

**Comment by redaJarir** [comment]
Hi, @anirban-mu , any news on the issue? 
I have the same issue as you for another library(pytimetk) (I'm on Mac OS Sonoma). The "import" doesn't work when  launched from RStudio with reticulate, but works fine when I use it directly in the terminal.

**Comment by anirban-mu** [comment]
Yep, looks like it's the same issue arising from the import of llvmlite.binding. Unfortunately, this is beyond my knowledge-base to debug.

```
> reticulate::repl_python()
Python 3.11.7 (/opt/homebrew/Cellar/python@3.11/3.11.7/bin/python3)
Reticulate 1.34.0 REPL -- A Python interpreter in R.
Enter 'exit' or 'quit' to exit the REPL and return to R.
>>> import pytimetk
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/ffi.py", line 134, in __getattr__
    return self._fntab[name]
           ~~~~~~~~~~~^^^^^^
KeyError: 'LLVMPY_AddSymbol'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/ffi.py", line 115, in _load_lib
    self._lib_handle = ctypes.CDLL(str(lib_path))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.7/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ctypes/__init__.py", line 376, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: dlopen(/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/libllvmlite.dylib, 0x0006): Library not loaded: @rpath/libz.1.dylib
  Referenced from: <F814708F-6874-3A38-AD06-6C06514419D4> /opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/libllvmlite.dylib
  Reason: tried: '/Library/Frameworks/R.framework/Resources/lib/libz.1.dylib' (no such file), '/opt/homebrew/Cellar/openjdk/21.0.1/libexec/openjdk.jdk/Contents/Home/lib/server/libz.1.dylib' (no such file)

<-- snip -- >
```

**Comment by anirban-mu** [comment]
Is there any chance that the handling of `DYLD_FALLBACK_LIBRARY_PATH` can be made user selectable in either RStudio or Reticulate? I.e., at startup, Reticulate either uses `DYLD_FALLBACK_LIBRARY_PATH` as is typical in R and therefore emulates a "Python in R environment" or as is typical in the interpreter given System Integrity Protection.

Totally understood if this is too corner case to work on or make changes for but I do think that in some cases, when considering Python through reticulate, people may expect and want an emulation of the interpreter while in others they may want to use both Python and R. Its just that it would be one thing if this issue occured with the Python interpreter but given that it is specific to Reticulate and some versions of RStudio, its difficult to get a handle on what is going on or how to fix it.

**Comment by MonkeySylvia** [comment]
Any updates on this - have been having the identical issue!

**Comment by kevinushey** [comment]
This may be resolved in the latest daily builds. We discovered that conda builds of R never set `LD_LIBRARY_PATH`, as described in https://github.com/rstudio/rstudio/issues/13184#issuecomment-1992709588. Unfortunately, this breaks RStudio as it relies on `LD_LIBRARY_PATH` being set to discover libraries required by R and their packages for conda installations.

We now have a workaround for this issue in the daily builds of RStudio at https://dailies.rstudio.com; it would be good to know if the issue still reproduces there.

**Comment by anirban-mu** [comment]
As of RStudio 2024.04.0 Build 663, I get the same error as earlier.

```
R Under development (unstable) (2024-03-10 r86083) -- "Unsuffered Consequences"
Copyright (C) 2024 The R Foundation for Statistical Computing
Platform: aarch64-apple-darwin23.3.0

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> 
> rm(list = ls())
> 
> # Are we running on the cloud?
> cloud <- FALSE
> 
> # Setup which python to use
> if (cloud) {
+   Sys.setenv(RETICULATE_PYTHON = "/opt/conda/bin/python3")
+   Sys.setenv(TENSORFLOW_PYTHON = "/opt/conda/bin/python3")
+   Sys.setenv(OMP_NUM_THREADS = toString(parallel::detectCores()))
+ } else {
+   Sys.setenv(RETICULATE_PYTHON = "/opt/Homebrew/Cellar/python@3.11/3.11.8/bin/python3.11")
+   Sys.setenv(TENSORFLOW_PYTHON = "/opt/Homebrew/Cellar/python@3.11/3.11.8/bin/python3.11")
+ }
> 
> library(reticulate)
> 
> reticulate::repl_python()
Python 3.11.8 (/opt/homebrew/Cellar/python@3.11/3.11.8/bin/python3.11)
Reticulate 1.35.0 REPL -- A Python interpreter in R.
Enter 'exit' or 'quit' to exit the REPL and return to R.
>>> import pandas as pd
>>> import random
>>> import re
>>> import time
>>> from openai import OpenAI
>>> import numba
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/ffi.py", line 136, in __getattr__
    return self._fntab[name]
           ~~~~~~~~~~~^^^^^^
KeyError: 'LLVMPY_AddSymbol'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/ffi.py", line 117, in _load_lib
    self._lib_handle = ctypes.CDLL(str(lib_path))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.8/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ctypes/__init__.py", line 376, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: dlopen(/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/libllvmlite.dylib, 0x0006): Library not loaded: /usr/local/lib/libc++.dylib
  Referenced from: <171BB73A-DA15-3323-B7A4-5FA86BBB3D28> /opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/libllvmlite.dylib
  Reason: tried: '/usr/local/lib/libc++.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/lib/libc++.dylib' (no such file), '/usr/local/lib/libc++.dylib' (no such file), '/Library/Frameworks/R.framework/Resources/lib/libc++.dylib' (no such file), '/opt/homebrew/Cellar/openjdk/21.0.2/libexec/openjdk.jdk/Contents/Home/lib/server/libc++.dylib' (no such file), '/var/folders/qd/7j99j7f123l170w8h2dr6yr80000gn/T/rstudio-fallback-library-path-1830315640/libc++.dylib' (no such file)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 119, in _find_and_load_hook
    return _run_hook(name, _hook)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 93, in _run_hook
    module = hook()
             ^^^^^^
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 117, in _hook
    return _find_and_load(name, import_)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/numba/__init__.py", line 70, in <module>
    from numba.core import config
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 119, in _find_and_load_hook
    return _run_hook(name, _hook)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 93, in _run_hook
    module = hook()
             ^^^^^^
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 117, in _hook
    return _find_and_load(name, import_)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/numba/core/config.py", line 17, in <module>
    import llvmlite.binding as ll
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 119, in _find_and_load_hook
    return _run_hook(name, _hook)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 93, in _run_hook
    module = hook()
             ^^^^^^
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 117, in _hook
    return _find_and_load(name, import_)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/__init__.py", line 4, in <module>
    from .dylib import *
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 119, in _find_and_load_hook
    return _run_hook(name, _hook)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 93, in _run_hook
    module = hook()
             ^^^^^^
  File "/Users/anirbanmukherjee/Library/R/arm64/4.4/library/reticulate/python/rpytools/loader.py", line 117, in _hook
    return _find_and_load(name, import_)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/dylib.py", line 36, in <module>
    ffi.lib.LLVMPY_AddSymbol.argtypes = [
    ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/ffi.py", line 139, in __getattr__
    cfn = getattr(self._lib, name)
                  ^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/ffi.py", line 131, in _lib
    self._load_lib()
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/ffi.py", line 125, in _load_lib
    raise OSError("Could not find/load shared object file") from e
OSError: Could not find/load shared object file
```

This is the only package where I have this error. I think I use about 30-50 Python packages across projects and the rest all work as expected.

**Comment by ronblum** [comment]
@anirban-mu Build 663 didn't have the fix yet. Today's build 668 will likely have it if you'd like to try it out, or wait for tomorrow's build.

**Comment by anirban-mu** [comment]
Build 668. Python interpreter:

```
anirbanmukherjee@Cosmo-93 ~ % /opt/Homebrew/Cellar/python@3.11/3.11.8/bin/python3.11
Python 3.11.8 (main, Feb  6 2024, 21:21:21) [Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numba
>>> 
```

Reticulate/R below. 

```
>>> import numba
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/ffi.py", line 136, in __getattr__
    return self._fntab[name]
           ~~~~~~~~~~~^^^^^^
KeyError: 'LLVMPY_AddSymbol'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/ffi.py", line 117, in _load_lib
    self._lib_handle = ctypes.CDLL(str(lib_path))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.8/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ctypes/__init__.py", line 376, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: dlopen(/opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/libllvmlite.dylib, 0x0006): Library not loaded: @rpath/libz.1.dylib
  Referenced from: <171BB73A-DA15-3323-B7A4-5FA86BBB3D28> /opt/homebrew/lib/python3.11/site-packages/llvmlite/binding/libllvmlite.dylib
  Reason: tried: '/Library/Frameworks/R.framework/Resources/lib/libz.1.dylib' (no such file), '/opt/homebrew/Cellar/openjdk/21.0.2/libexec/openjdk.jdk/Contents/Home/lib/server/libz.1.dylib' (no such file), '/var/folders/qd/7j99j7f123l170w8h2dr6yr80000gn/T/rstudio-fallback-library-path-757458580/libz.1.dylib' (no such file)

<snip>
```

**Comment by ronblum** [comment]
@kevinushey Looks like this is still an issue.

**Comment by kevinushey** [comment]
Notes from trying to reproduce. I have:

```
$ csrutil status
System Integrity Protection status: enabled.
```

And, in R:

```
> packageVersion("reticulate")
[1] '1.35.0'
```

I then installed python 3.11.6 with:

```
> reticulate::install_python(version = "3.11.6")
Cloning into '/Users/kevin/.pyenv/plugins/pyenv-update'...
+ /opt/homebrew/bin/pyenv update
+ /opt/homebrew/bin/pyenv install --skip-existing 3.11.6
python-build: use openssl@3 from homebrew
python-build: use readline from homebrew
Downloading Python-3.11.6.tar.xz...
-> https://www.python.org/ftp/python/3.11.6/Python-3.11.6.tar.xz
Installing Python-3.11.6...
python-build: use tcl-tk from homebrew
python-build: use readline from homebrew
python-build: use ncurses from homebrew
python-build: use zlib from xcode sdk
Installed Python-3.11.6 to /Users/kevin/.pyenv/versions/3.11.6
[1] "/Users/kevin/.pyenv/versions/3.11.6/bin/python3.11"
```

Then, I restarted my R session and activated that Python installation:

```
python <- "/Users/kevin/.pyenv/versions/3.11.6/bin/python3.11"
Sys.setenv(RETICULATE_PYTHON = python)
Sys.setenv(TENSORFLOW_PYTHON = python)
library(reticulate)
```

Then, install the required Python packages:

```
kevin@MBP-P2MQ:~/.pyenv/versions/3.11.6/bin
$ ./python -m pip install numpy pandas umap
Collecting numpy
  Obtaining dependency information for numpy from https://files.pythonhosted.org/packages/1a/2e/151484f49fd03944c4a3ad9c418ed193cfd02724e138ac8a9505d056c582/numpy-1.26.4-cp311-cp311-macosx_11_0_arm64.whl.metadata
  Downloading numpy-1.26.4-cp311-cp311-macosx_11_0_arm64.whl.metadata (114 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.8/114.8 kB 6.8 MB/s eta 0:00:00
Collecting pandas
  Obtaining dependency information for pandas from https://files.pythonhosted.org/packages/a5/78/1d859bfb619c067e3353ed079248ae9532c105c4e018fa9a776d04b34572/pandas-2.2.1-cp311-cp311-macosx_11_0_arm64.whl.metadata
  Downloading pandas-2.2.1-cp311-cp311-macosx_11_0_arm64.whl.metadata (19 kB)
Collecting umap
  Using cached umap-0.1.1.tar.gz (3.2 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting python-dateutil>=2.8.2 (from pandas)
  Obtaining dependency information for python-dateutil>=2.8.2 from https://files.pythonhosted.org/packages/ec/57/56b9bcc3c9c6a792fcbaf139543cee77261f3651ca9da0c93f5c1221264b/python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata
  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
Collecting pytz>=2020.1 (from pandas)
  Obtaining dependency information for pytz>=2020.1 from https://files.pythonhosted.org/packages/9c/3d/a121f284241f08268b21359bd425f7d4825cffc5ac5cd0e1b3d82ffd2b10/pytz-2024.1-py2.py3-none-any.whl.metadata
  Downloading pytz-2024.1-py2.py3-none-any.whl.metadata (22 kB)
Collecting tzdata>=2022.7 (from pandas)
  Obtaining dependency information for tzdata>=2022.7 from https://files.pythonhosted.org/packages/65/58/f9c9e6be752e9fcb8b6a0ee9fb87e6e7a1f6bcab2cdc73f02bb7ba91ada0/tzdata-2024.1-py2.py3-none-any.whl.metadata
  Downloading tzdata-2024.1-py2.py3-none-any.whl.metadata (1.4 kB)
Collecting six>=1.5 (from python-dateutil>=2.8.2->pandas)
  Obtaining dependency information for six>=1.5 from https://files.pythonhosted.org/packages/d9/5a/e7c31adbe875f2abbb91bd84cf2dc52d792b5a01506781dbcf25c91daf11/six-1.16.0-py2.py3-none-any.whl.metadata
  Downloading six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Downloading numpy-1.26.4-cp311-cp311-macosx_11_0_arm64.whl (14.0 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.0/14.0 MB 77.7 MB/s eta 0:00:00
Downloading pandas-2.2.1-cp311-cp311-macosx_11_0_arm64.whl (11.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.3/11.3 MB 83.0 MB/s eta 0:00:00
Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 229.9/229.9 kB 26.6 MB/s eta 0:00:00
Downloading pytz-2024.1-py2.py3-none-any.whl (505 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 505.5/505.5 kB 35.6 MB/s eta 0:00:00
Downloading tzdata-2024.1-py2.py3-none-any.whl (345 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.4/345.4 kB 30.0 MB/s eta 0:00:00
Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Building wheels for collected packages: umap
  Building wheel for umap (pyproject.toml) ... done
  Created wheel for umap: filename=umap-0.1.1-py3-none-any.whl size=3543 sha256=3a21cf1c7a0d47853c4a4a94b6e29b8d4c8dfb51eb7ba183fac8d2d01903d338
  Stored in directory: /Users/kevin/Library/Caches/pip/wheels/82/d8/73/e9eb3334baaad795ff0278363ff1aca7568bdf2793e452a527
Successfully built umap
Installing collected packages: umap, pytz, tzdata, six, numpy, python-dateutil, pandas
Successfully installed numpy-1.26.4 pandas-2.2.1 python-dateutil-2.9.0.post0 pytz-2024.1 six-1.16.0 tzdata-2024.1 umap-0.1.1

[notice] A new release of pip is available: 23.2.1 -> 24.0
[notice] To update, run: /Users/kevin/.pyenv/versions/3.11.6/bin/python -m pip install --upgrade pip
```

Then, in a fresh R session:

```
> python <- "/Users/kevin/.pyenv/versions/3.11.6/bin/python3.11"
> Sys.setenv(RETICULATE_PYTHON = python)
> Sys.setenv(TENSORFLOW_PYTHON = python)
> library(reticulate)
> 
> py_config()
python:         /Users/kevin/.pyenv/versions/3.11.6/bin/python3.11
libpython:      /Users/kevin/.pyenv/versions/3.11.6/lib/libpython3.11.dylib
pythonhome:     /Users/kevin/.pyenv/versions/3.11.6:/Users/kevin/.pyenv/versions/3.11.6
version:        3.11.6 (main, Apr  1 2024, 10:21:30) [Clang 15.0.0 (clang-1500.3.9.4)]
numpy:          /Users/kevin/.pyenv/versions/3.11.6/lib/python3.11/site-packages/numpy
numpy_version:  1.26.4

NOTE: Python version was forced by RETICULATE_PYTHON
> repl_python(input = c(
+    "import numpy as np",
+    "import pandas as pd",
+    "import umap"
+ ))
Python 3.11.6 (/Users/kevin/.pyenv/versions/3.11.6/bin/python3.11)
Reticulate 1.35.0 REPL -- A Python interpreter in R.
Enter 'exit' or 'quit' to exit the REPL and return to R.
>>>  import numpy as np
>>>  import pandas as pd
>>>  import umap
```

And everything appears to be fine.

```
> sessionInfo()
R version 4.3.3 (2024-02-29)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.4.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Los_Angeles
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] reticulate_1.35.0

loaded via a namespace (and not attached):
 [1] utf8_1.2.4        Matrix_1.6-5      lattice_0.22-5    glue_1.7.0        png_0.1-8         lifecycle_1.0.4   cli_3.6.2        
 [8] fansi_1.0.6       grid_4.3.3        vctrs_0.6.5       compiler_4.3.3    rstudioapi_0.16.0 tools_4.3.3       pillar_1.9.0     
[15] Rcpp_1.0.12       rlang_1.1.3       jsonlite_1.8.8 
```

**Comment by anirban-mu** [comment]
One difference is that I installed Python (specifically Python@3.11) using home brew (i.e., `brew install python@3.11`). No specific reason beyond that if all other libraries are from brew, I figured probably best to also get Python from it? But I see in your code:

```
Downloading Python-3.11.6.tar.xz...
-> https://www.python.org/ftp/python/3.11.6/Python-3.11.6.tar.xz
Installing Python-3.11.6...
```

Let me know if I can provide further information. I don't have any ideas but this is far from my field of expertise as I am only an end user.

**Comment by anirban-mu** [comment]
Given @kevinushey's last message, I had a hunch that this may be related to installing Python from a brew binary vs. from python.org. So I did `brew reinstall --build-from-source python@3.11`. This solved the issue on my system for both R on Terminal and RStudio whereby I know get 

```
> library(reticulate)
> 
> reticulate::repl_python()
Python 3.11.9 (/opt/homebrew/Cellar/python@3.11/3.11.9/bin/python3.11)
Reticulate 1.35.0 REPL -- A Python interpreter in R.
Enter 'exit' or 'quit' to exit the REPL and return to R.
>>> # Import necessary libraries
>>> import umap
>>>
```

I did not reinstall R or umap. I only reinstalled Python 3.11 from source using Homebrew (installing Python 3.11.9 from brew binary did not resolve, it required installing from source). Also, I am now using RStudio 2023.12.1 Build 402 and not the daily.
